src/veritysetup/veritysetup.c

```diff
@@ -67,6 +79,25 @@ static int looks_like_roothashsig(const char *option) {
         return 0;
 }
 
+static int parse_block_size(const char *t, uint64_t *size) {
+        uint64_t u;
+        int r;
+
+        r = parse_size(t, 1024, &u);
+        if (r < 0)
+                return r;
+
+        if (u < 512 || u > (512 * 1024))
+                return -ERANGE;
+
+        if (u % 512 || (u & (u-1)))
```

Comment:
The code was originally taken from [cryptsetup](https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/verity/verity.h#L28-L29) and I have split it up to make it more digest to read.

I have no idea of what that code is supposed to test; the git [history](https://gitlab.com/cryptsetup/cryptsetup/-/blame/main/lib/verity/verity.h#L28) does not say much about that checking.

In general I trust the underlying library for checking the given inputs:
- this avoids checking them twice; first in `systemd-veritysetup` then in `cryptsetup`
- and we can remove the "duplicated" checking-code in systemd's `veritysetup.c` (i.e. remove the `parse_block_size()` function and the checks in function `parse_options()`). Note: this applies to the `format=` checking as well; `0` and `1` allowed for now.

In my mind, `systemd-veritysetup` is just a simple wrapper around the `cryptsetup` library. In the situation of a new verity format/hash is released by the kernel in the future (i.e. `format=2`) we do not have to update the code in systemd (if the check for format/hash is not performed in systemd). We just need a new version of the `cryptsetup` library installed on the system with the following format/hash [checking](https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/setup.c#L2033-2036) updated.

However, this is just a thinking :)

Reply:
So, I decided to sit down and add a proper ISPOWEROF2() macro to our codebase, see #18789 Can you rebase on that, and use it for the right hand check here?

Reply:
So, we typically want to validate things roughly ourselves, before accepting it, so that we can generate nice error messages. handing everything off to libcryptsetup doesn't really allow that...

Reply:
Ok, I will do.

Reply:
done.